ambry-store/src/main/java/com.github.ambry.store/IndexValue.java

```diff
@@ -13,114 +13,322 @@
+          throw new IllegalArgumentException("Invalid buffer size for version 0");
+        }
+        size = value.getLong();
+        offset = new Offset(logSegmentName, value.getLong());
+        flags = value.get();
+        expiresAtMs = value.getLong();
+        originalMessageOffset = value.getLong();
+        operationTimeInMs = (int) Utils.Infinite_Time;
+        serviceId = SERVICE_CONTAINER_ID_DEFAULT_VALUE;
+        containerId = SERVICE_CONTAINER_ID_DEFAULT_VALUE;
+        break;
+      case PersistentIndex.VERSION_1:
+        if (value.capacity() != INDEX_VALUE_SIZE_IN_BYTES_V1) {
+          throw new IllegalArgumentException("Invalid buffer size for version 1");
+        }
+        size = value.getLong();
+        offset = new Offset(logSegmentName, value.getLong());
+        flags = value.get();
+        int expiresAt = value.getInt();
+        expiresAtMs = expiresAt != Utils.Infinite_Time ? (expiresAt * Time.MsPerSec) : Utils.Infinite_Time;
+        originalMessageOffset = value.getLong();
+        int operationTimeInSecs = value.getInt();
+        operationTimeInMs =
+            operationTimeInSecs != Utils.Infinite_Time ? (operationTimeInSecs * Time.MsPerSec) : Utils.Infinite_Time;
+        serviceId = value.getShort();
+        containerId = value.getShort();
+        break;
+      default:
+        throw new IllegalArgumentException("Unsupported version " + version + " passed in for IndexValue ");
     }
-    this.value = value;
-    offset = new Offset(logSegmentName, value.getLong(Blob_Size_In_Bytes));
   }
 
-  IndexValue(long size, Offset offset, byte flags, long expiresAtMs) {
-    this(size, offset, flags, expiresAtMs, offset.getOffset());
+  /**
+   * Constructs IndexValue based on the args passed
+   * @param size the size of the blob that this index value refers to
+   * @param offset the {@link Offset} in the {@link Log} where the blob that this index value refers to resides
+   * @param expiresAtMs the expiration time in ms at which the blob expires
+   */
+  IndexValue(long size, Offset offset, long expiresAtMs) {
+    this(size, offset, FLAGS_DEFAULT_VALUE, expiresAtMs, offset.getOffset(), (int) Utils.Infinite_Time,
+        SERVICE_CONTAINER_ID_DEFAULT_VALUE, SERVICE_CONTAINER_ID_DEFAULT_VALUE);
   }
 
-  IndexValue(long size, Offset offset, long expiresAtMs) {
-    this(size, offset, (byte) 0, expiresAtMs);
+  /**
+   * Constructs IndexValue based on the args passed
+   * @param size the size of the blob that this index value refers to
+   * @param offset the {@link Offset} in the {@link Log} where the blob that this index value refers to resides
+   * @param flags the {@link Flags} that needs to be set for the Index Value
+   * @param expiresAtMs the expiration time in ms at which the blob expires
+   * @param operationTimeInMs operation time ins ms of the entry in secs
+   */
+  IndexValue(long size, Offset offset, byte flags, long expiresAtMs, long operationTimeInMs) {
```

Comment:
Once we add changes to ensure we add in the correct operationTime, this constructor will be in use. For the serviceId, containerId changes to come, its gonna take much more time compared to operationTime changes